File: javascript/sdk/ussd.js
/**
 * USSD module - USSD Sender and Receiver classes
 * Equivalent to PHP UssdSender and UssdReceiver classes in sdk_file.php
 */

const Core = require('./core');

// ─── Custom Exception ───────────────────────────────────────────────
class UssdException extends Error {
  constructor(message, code, response = null) {
    super(message);
    this.name = 'UssdException';
    this.statusMessage = message;
    this.code = code;
    this.rawResponse = response;
  }

  getStatusCode() { return this.code; }
  getStatusMessage() { return this.statusMessage; }
  getRawResponse() { return this.rawResponse; }
}

// ─── USSD Receiver ──────────────────────────────────────────────────
class UssdReceiver {
  /**
   * Parse incoming USSD request JSON
   * @param {object} data - Parsed JSON body from incoming request
   */
  constructor(data) {
    if (!data || (!data.sourceAddress && !data.message)) {
      throw new Error('Some of the required parameters are not provided');
    }

    this.sourceAddress = data.sourceAddress || null;
    this.message = data.message || null;
    this.requestId = data.requestId || null;
    this.applicationId = data.applicationId || null;
    this.encoding = data.encoding || null;
    this.version = data.version || null;
    this.sessionId = data.sessionId || null;
    this.ussdOperation = data.ussdOperation || null;
    this.thejson = data;

    this.response = { statusCode: 'S1000', statusDetail: 'Success' };
  }

  getJson() { return this.thejson; }
  getAddress() { return this.sourceAddress; }
  getMessage() { return this.message; }
  getRequestID() { return this.requestId; }
  getApplicationId() { return this.applicationId; }
  getEncoding() { return this.encoding; }
  getVersion() { return this.version; }
  getSessionId() { return this.sessionId; }
  getUssdOperation() { return this.ussdOperation; }
  getResponse() { return this.response; }
}

// ─── USSD Sender ────────────────────────────────────────────────────
class UssdSender extends Core {
  /**
   * @param {string} serverURL - BDApps USSD send URL
   * @param {string} applicationId - Application ID
   * @param {string} password - Application password
   */
  constructor(serverURL, applicationId, password) {
    super();
    this.serverURL = serverURL;
    this.applicationId = applicationId;
    this.password = password;
  }

  /**
   * Send USSD message
   * @param {string} sessionId - Session identifier
   * @param {string} message - Message content
   * @param {string} destinationAddress - Destination phone number
   * @param {string} ussdOperation - USSD operation type (default: 'mo-cont')
   * @returns {Promise<object>} Response from BDApps
   */
  async ussd(sessionId, message, destinationAddress, ussdOperation = 'mo-cont') {
    if (!destinationAddress) {
      throw new Error('address should be a string or an array of strings');
    }
    return this._ussdMany(message, sessionId, ussdOperation, destinationAddress);
  }

  /** @private */
  async _ussdMany(message, sessionId, ussdOperation, destinationAddress) {
    const payload = {
      applicationId: this.applicationId,
      password: this.password,
      message,
      destinationAddress,
      sessionId,
      ussdOperation,
      encoding: '440',
    };

    return this.sendRequest(payload, this.serverURL);
  }
}

module.exports = { UssdSender, UssdReceiver, UssdException };

File: python/sdk/ussd.py
"""
USSD module - USSD Sender and Receiver classes.
Equivalent to PHP UssdSender and UssdReceiver classes in sdk_file.php.
"""

from typing import Optional

from .core import Core
from .exceptions import BDAppsException


class UssdException(BDAppsException):
    """Exception for USSD service errors."""
    pass


class UssdReceiver:
    """
    Parse incoming USSD request JSON.
    Equivalent to PHP UssdReceiver class.
    """

    def __init__(self, data: dict):
        if not data or ("sourceAddress" not in data and "message" not in data):
            raise ValueError("Some of the required parameters are not provided")

        self.source_address = data.get("sourceAddress")
        self.message = data.get("message")
        self.request_id = data.get("requestId")
        self.application_id = data.get("applicationId")
        self.encoding = data.get("encoding")
        self.version = data.get("version")
        self.session_id = data.get("sessionId")
        self.ussd_operation = data.get("ussdOperation")
        self.thejson = data
        self.response = {"statusCode": "S1000", "statusDetail": "Success"}

    def get_json(self) -> dict:
        return self.thejson

    def get_address(self) -> Optional[str]:
        return self.source_address

    def get_message(self) -> Optional[str]:
        return self.message

    def get_request_id(self) -> Optional[str]:
        return self.request_id

    def get_application_id(self) -> Optional[str]:
        return self.application_id

    def get_encoding(self) -> Optional[str]:
        return self.encoding

    def get_version(self) -> Optional[str]:
        return self.version

    def get_session_id(self) -> Optional[str]:
        return self.session_id

    def get_ussd_operation(self) -> Optional[str]:
        return self.ussd_operation

    def get_response(self) -> dict:
        return self.response


class UssdSender(Core):
    """
    Send USSD messages.
    Equivalent to PHP UssdSender class.
    """

    def __init__(self, server_url: str, application_id: str, password: str):
        super().__init__()
        self.server_url = server_url
        self.application_id = application_id
        self.password = password

    async def ussd(
        self,
        session_id: str,
        message: str,
        destination_address: str,
        ussd_operation: str = "mo-cont",
    ) -> dict:
        """
        Send USSD message.

        Args:
            session_id: Session identifier.
            message: Message content.
            destination_address: Destination phone number.
            ussd_operation: USSD operation type (default: 'mo-cont').

        Returns:
            Response from BDApps.
        """
        if not destination_address:
            raise ValueError("address should be a string or an array of strings")

        return await self._ussd_many(
            message, session_id, ussd_operation, destination_address
        )

    async def _ussd_many(
        self,
        message: str,
        session_id: str,
        ussd_operation: str,
        destination_address: str,
    ) -> dict:
        payload = {
            "applicationId": self.application_id,
            "password": self.password,
            "message": message,
            "destinationAddress": destination_address,
            "sessionId": session_id,
            "ussdOperation": ussd_operation,
            "encoding": "440",
        }

        return await self.send_request(payload, self.server_url)
